staticaddr/deposit/fsm.go

```diff
@@ -0,0 +1,303 @@
+package deposit
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightninglabs/lndclient"
+	"github.com/lightninglabs/loop/fsm"
+	"github.com/lightninglabs/loop/staticaddr"
+	"github.com/lightninglabs/loop/staticaddr/address"
+	"github.com/lightninglabs/loop/staticaddr/script"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/keychain"
+)
+
+const (
+	DefaultObserverSize = 20
+)
+
+var (
+	ErrProtocolVersionNotSupported = errors.New("protocol version not " +
+		"supported")
+)
+
+// States.
+var (
+	Deposited = fsm.StateType("Deposited")
+
+	PublishExpiredDeposit = fsm.StateType("PublishExpiredDeposit")
+
+	WaitForExpirySweep = fsm.StateType("WaitForExpirySweep")
+
+	Expired = fsm.StateType("Expired")
+
+	Failed = fsm.StateType("Failed")
+)
+
+// Events.
+var (
+	OnStart           = fsm.EventType("OnStart")
+	OnExpiry          = fsm.EventType("OnExpiry")
+	OnExpiryPublished = fsm.EventType("OnExpiryPublished")
+	OnExpirySwept     = fsm.EventType("OnExpirySwept")
+	OnRecover         = fsm.EventType("OnRecover")
+)
+
+// FSM is the state machine that handles the instant out.
+type FSM struct {
+	*fsm.StateMachine
+
+	cfg *ManagerConfig
+
+	deposit *Deposit
+
+	params *address.Parameters
+
+	address *script.StaticAddress
+
+	ctx context.Context
+
+	blockNtfnChan chan uint32
+
+	finalizedDepositChan chan wire.OutPoint
+}
+
+// NewFSM creates a new state machine that can action on all static address
+// feature requests.
+func NewFSM(ctx context.Context, deposit *Deposit, cfg *ManagerConfig,
+	finalizedDepositChan chan wire.OutPoint,
+	recoverStateMachine bool) (*FSM, error) {
+
+	params, err := cfg.AddressManager.GetStaticAddressParameters(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get static address "+
+			"parameters: %v", err)
+	}
+
+	address, err := cfg.AddressManager.GetStaticAddress(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get static address: %v", err)
+	}
+
+	depoFsm := &FSM{
+		cfg:                  cfg,
+		deposit:              deposit,
+		params:               params,
+		address:              address,
+		ctx:                  ctx,
+		blockNtfnChan:        make(chan uint32),
+		finalizedDepositChan: finalizedDepositChan,
+	}
+
+	depositStates := depoFsm.DepositStatesV0()
+	switch params.ProtocolVersion {
```

Comment:
nit: this check could go just below where we call `GetStaticAddressParameters()`.

Reply:
I think it has to stay here since if there'd be a `V1` state machine we need the `depoFsm` struct to retrieve those states.